api/flask-app/app.py

```diff
@@ -11,64 +34,358 @@ def index():
+            cursor.execute(sql, [cursor.lastrowid])
+            inserted_row = cursor.fetchone()
+        return jsonify(inserted_row), 201
+
+    else:
+        # Unsupported HTTP method
+        return Response(status=405)
+
+
+@app.route("/users/<id>", methods=["GET", "PUT", "DELETE"])
+def users_by_id(id):
+    db = get_db()
+    row = get_by_id(db.cursor(), "User", id)
+    if not row:
+        return Response(status=404)
+
+    if request.method == "GET":
+        return jsonify(row)
+
     elif request.method == "PUT":
-        # prepare SQL query from api call params
-        # send query
-        # return something (could be updated rows or no contents)
-        return Response(status=200)
+        schema = UserSchema()
+        try:
+            user = schema.load(request.get_json())
+        except (ValidationError, TypeError):
+            return Response(status=400)
+        user_json = schema.dump(user)
+
+        with db.cursor() as cursor:
+            sql = "UPDATE User SET name=%(name)s, email=%(email)s, " \
+                  "bio=%(bio)s WHERE id=%(id)s"
+            args = user_json
+            args["id"] = id
+            cursor.execute(sql, args)
+
+        db.commit()
+        row = get_by_id(db.cursor(), "User", id)
+        return jsonify(row)
+
+    elif request.method == "DELETE":
+        delete_by_id(db.cursor(), "User", id)
+        db.commit()
+        return Response(status=204)
+
+    else:
+        # Unsupported HTTP method
+        return Response(status=405)
+
+
+@app.route("/users/<id>/followers", methods=["GET"])
+def followers(id):
+    db = get_db()
+    user = get_by_id(db.cursor(), "User", id)
+    if not user:
+        return Response(status=404)
+    with db.cursor() as cursor:
+        sql = "SELECT follower_id FROM Follow WHERE following_id=%s"
+        cursor.execute(sql, [id])
+        follower_ids = cursor.fetchall()
+
+        sql = "SELECT * FROM User WHERE id IN (%s)"
+        cursor.execute(
+            sql,
+            [', '.join(map(lambda x: str(x["follower_id"]), follower_ids))]
+        )
+        followers = cursor.fetchall()
+        return jsonify(followers)
+
+
+@app.route("/users/<id>/following", methods=["GET"])
+def following(id):
+    db = get_db()
+    user = get_by_id(db.cursor(), "User", id)
+    if not user:
+        return Response(status=404)
+    following_str = get_following_str(db.cursor(), id)
+
+    with db.cursor() as cursor:
+        sql = "SELECT * FROM User WHERE id IN (%s)"
+        cursor.execute(sql, [following_str])
+        following = cursor.fetchall()
+    return jsonify(following)
+
+
+@app.route("/users/<id>/feed", methods=["GET"])
+def feed(id):
+    db = get_db()
+    user = get_by_id(db.cursor(), "User", id)
+    if not user:
+        return Response(status=404)
+    following_str = get_following_str(db.cursor(), id)
+    with db.cursor() as cursor:
+        sql = "SELECT * FROM Message WHERE poster in (%s)"
```

Comment:
Does this get all messages from one user?

Reply:
This should get all messages from the users that a user is following. Here, `%s` will evaluate to a comma-separated string containing all the following user ids for a specific user. So if a user is following two users with ids `1` and `2`, the sql query will be `SELECT * FROM Message WHERE poster in (1, 2)`